### DISCUSSION TITLE
How to pass my .kubeconfig file with Kubernetes certificates, host name etc to controllers?

### DISCUSSION THREAD

**Discussion by vvitkovsky** [dotnet/dotnet-operator-sdk#544] [Q&A] [answered]
Hello,

I started to investigate this project, but I can't find any example how I can pass my own .kubeconfig file for a Kubernetes client that should be used in controllers, could you please advise me? 

Best regards, 
Victor

**Comment by buehler** [comment, marked as answer]
Hey Victor

Thank you for your interesst in this repo :)
The .kubeconfig used is the default (config in your userprofile). This is sufficient for most of the usecases (e.g running the operator in kubernetes).

Currently there is no possibility to setup your own kubernetes client (with an "official public api"). You can however (as workaround) use:
```
builder.Services.AddKubernetesOperator();
// prosa code, don't know the exact calls:
builder.Services.Remove<IKubernetesClient>;
builder.Services.AddSingleton<IKubernetesClient>(new KubernetesClient(...));
```

**Reply by vvitkovsky** [reply]
Hi Christoph, 

Thank you for your answer.
Btw, you can easily pass this kubeconfig path over OperatorSettings:
OperatorBuilder.cs:
```
internal IOperatorBuilder AddOperatorBase(OperatorSettings settings) 
{
...
Services.AddSingleton<IKubernetesClient, KubernetesClient.KubernetesClient>();
...
}
```
But to continue I can use workaround that you have mentioned.

Thanks,
Victor

**Reply by vvitkovsky** [reply]
Btw, I have found a bit more elegant workaround:
KubeOps create own client based on .net Kubernetes client with default config:

```
public KubernetesClient()
        : this(KubernetesClientConfiguration.BuildDefaultConfig())
{
}
```

but official .net Kubernetes client uses "KUBECONFIG" environment variable first to check config path, then it looks for default home folder. So, we can easily override this environment variable before call to 

`builder.Services.AddKubernetesOperator()`

like this:

```
var path = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "my.kubeconfig");
Environment.SetEnvironmentVariable("KUBECONFIG", path);
```

now it uses your kubeconfig file in KubeOps

Best regards,
Victor

**Reply by buehler** [reply]
Oh yes, it sliped my mind, that works as well :-)
